runtime/jsMain/src/kotlinx/serialization/DynamicObjectSerializer.kt

```diff
@@ -0,0 +1,182 @@
+        }
+        return true
+    }
+
+    override fun encodeValue(value: Any) {
+        if (currentElementIsMapKey) {
+            currentName = value.toString()
+        } else {
+            current.jsObject[currentName] = value
+        }
+    }
+
+    override fun encodeChar(value: Char) {
+        encodeValue(value.toString())
+    }
+
+    override fun encodeNull() {
+        if (currentElementIsMapKey) {
+            currentName = null
+        } else {
+            if (encodeNullAsUndefined) return // omit element
+
+            current.jsObject[currentName] = null
+        }
+    }
+
+    override fun encodeEnum(enumDescription: SerialDescriptor, ordinal: Int) {
+        encodeValue(enumDescription.getElementName(ordinal))
+    }
+
+    override fun shouldEncodeElementDefault(desc: SerialDescriptor, index: Int) = encodeDefaults
+
+    fun enterNode(jsObject: dynamic, writeMode: WriteMode) {
+        val child = Node(writeMode, jsObject)
+        child.parent = current
+        current = child
+    }
+
+    fun exitNode() {
+        current = current.parent
+        currentElementIsMapKey = false
+    }
+
+    override fun beginStructure(desc: SerialDescriptor, vararg typeParams: KSerializer<*>): CompositeEncoder {
+        // we currently do not structures as map key
+        if (currentElementIsMapKey) {
+            throw IllegalArgumentException(
+                "Value of type ${desc.name} can't be used in json as map key. " +
+                        "It should have either primitive or enum kind, but its kind is ${desc.kind}.\n"
+            )
+        }
+
+        val newMode = selectMode(desc)
+        if (result === NoOutputMark) {
+
+            result = newChild(newMode)
+            current = Node(newMode, result)
+            current.parent = current
+        } else {
+            val child = newChild(newMode)
+            current.jsObject[currentName] = child
+            enterNode(child, newMode)
+        }
+
+        current.index = 0
+        return this
+    }
+
+    fun newChild(writeMode: WriteMode) = when (writeMode) {
+        WriteMode.OBJ, WriteMode.MAP -> js(BEGIN_OBJ.toString() + END_OBJ)
+        WriteMode.LIST -> js(BEGIN_LIST.toString() + END_LIST)
+    }
+
+    override fun endStructure(desc: SerialDescriptor) {
+        exitNode()
+    }
+
+    fun selectMode(desc: SerialDescriptor) =
+        when (desc.kind) {
+            StructureKind.CLASS, UnionKind.OBJECT -> WriteMode.OBJ
+            StructureKind.LIST, is PolymorphicKind -> WriteMode.LIST
+            StructureKind.MAP -> WriteMode.MAP
+            is PrimitiveKind, UnionKind.ENUM_KIND -> {
+                // the two cases are handled in DynamicObjectSerializer. But compiler does not know
+                throw IllegalArgumentException("DynamicObjectSerializer does not support serializing singular values of primitive or enum type.")
+            }
+        }
+
+    override fun toString(): String {
```

Comment:
Since it is useful only as a debug information while developing, I suggest removing this before merging.